llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -2821,6 +2821,38 @@ SDValue DAGCombiner::visitADDLike(SDNode *N) {
   return SDValue();
 }
 
+// Attempt to form ext(avgflooru(A, B)) from add(and(A, B), lshr(xor(A, B), 1))
+static SDValue combineFixedwidthToAVG(SDNode *N, SelectionDAG &DAG) {
+  assert(N->getOpcode() == ISD::ADD && "ADD node is required here");
+  SDValue And = N->getOperand(0);
+  SDValue Lshr = N->getOperand(1);
+  if (And.getOpcode() != ISD::AND || Lshr.getOpcode() != ISD::SRL)
```

Comment:
Ideally the operands of the `Add` would be treated commutatively - either could be the `and` and either could be the `srl`. Same for the `and`/`xor` args below.

Reply:
Using sd_match would handle commutative matching for us

Reply:
This is the kind of thing I had in mind:
```cpp
  if (TLI.isOperationLegal(ISD::AVGFLOORU, VT)) {
    SDValue A, B;
    if (sd_match(N, m_Add(m_And(m_Value(A), m_Value(B)),
                          m_Srl(m_Xor(m_Deferred(A), m_Deferred(B)),
                                m_SpecificInt(1))))) {
      return DAG.getNode(ISD::AVGFLOORU, DL, VT, A, B);
    }
  }
```

Reply:
But since `m_And` and `m_Xor` are not available in `SDPatternMatch`, we will have to use `llvm::PatternMatch` for them. But I don't think if this will work. This is what I get finally and it crashes the build process. 

```c
  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
  SDValue N0 = N->getOperand(0);
  EVT VT = N0.getValueType();
  SDLoc DL(N);
  if (TLI.isOperationLegal(ISD::AVGFLOORU, VT)) {
    SDValue A, B;
    if (sd_match(N, llvm::SDPatternMatch::m_Add(llvm::PatternMatch::m_And(m_Value(A), m_Value(B)),
                          llvm::SDPatternMatch::m_Srl(llvm::PatternMatch::m_Xor(m_Deferred(A), m_Deferred(B)),
                                m_SpecificInt(1))))) {
      return DAG.getNode(ISD::AVGFLOORU, DL, VT, A, B);
    }
  }
```

Reply:
I have added the code for commutative matching, but not using the method as suggested by @RKSimon.

Reply:
I'll get m_And / mXor / m_Or added shortly

Reply:
Right, sure thing.

Reply:
Modified the code accordingly.